File: Search2DMatrix.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        if (matrix.Length == 1) {
            if (matrix[0].Length == 0) {
                return false;
            }
        }
        for (int i = 0; i < matrix.Length; i++) {
            if (matrix[i][0] == target) {
                return true;
            }
            if (matrix[i][0] < target && matrix[i][matrix[i].Length - 1] > target) {
                Console.WriteLine($"{matrix[i][0]}, {matrix[i][matrix[i].Length - 1]}");
                int left = 0;
                int right = matrix[i].Length - 1;
                while (left <= right) {
                    int mid = left + (right - left) / 2;
                    Console.WriteLine($"{matrix[i][mid]}");
                    if (matrix[i][mid] == target) {
                        return true;
                    } else if (matrix[i][mid] > target) {
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
            } else {
                if (matrix[i][0] == target) {
                    return true;
                }
                if (matrix[i][matrix[i].Length - 1] == target) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: Search2DMatrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        if matrix.is_empty() || matrix[0].is_empty() {
            return false;
        }

        let m: i32 = matrix.len() as i32;
        let n: i32 = matrix[0].len() as i32;
        let mut left: i32 = 0;
        let mut right: i32 = (m * n) - 1;

        while left <= right {
            let mid: i32 = left + (right - left) / 2;
            let row: i32 = mid / n;
            let col: i32 = mid % n;

            if matrix[row as usize][col as usize] == target {
                return true;
            } else if matrix[row as usize][col as usize] > target {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        false
    }
}